### DISCUSSION TITLE
Using Registration Matrices from Flirt

### DISCUSSION THREAD

**Discussion by cascam07** [simnibs/simnibs#156] [7.Creating head models] [answered]
Our lab works with infant brain injury MRI data, and I am currently trying to build head models for some of out participants. I haven't had much success with the registrations built into Charm, which isn't terribly surprising, most registration tools calibrated to healthy adult data fail on lesioned infant data. We have, however, had success using the Flirt tool from FSL for registration, so I'd like to use the transformation matrices from Flirt to perform the template registration in Charm. I've tried using the --usetransform flag to provide an affine Flirt matrix I generated by registering the template (C:\SimNIBS\simnibs_env\Lib\site-packages\simnibs\segmentation\atlases\charm_atlas_mni) to the T1.

e.g.
`charm Test T1.nii.gz T2.nii.gz  --usetransform flirtmat`

When I run Charm this way, however, the registration comes out wrong. The scaling seems right but the template is translated off from the T1. Because of privacy concerns, I can't share our infant data here, but I have replicated the issue using the Ernie dataset.

Here is ernie in red and the template in blue, prior to any registration.
![image](https://github.com/simnibs/simnibs/assets/5588390/52f4f34e-2f40-48cd-903d-37443c9eacd3)

I confirmed that the registration matrix works when I load it in Fsleyes, so I'm pretty confident the matrix is reasonable. 
![image](https://github.com/simnibs/simnibs/assets/5588390/ff3ecebf-db40-4c6e-9fde-ef19c9510dfa)

But when I use this same registration matrix with Charm, this is the registration I get.
![image](https://github.com/simnibs/simnibs/assets/5588390/737db1b7-bcbf-4259-ae0a-112147402ae5)

I would appreciate any guidance you can give on using a previously determined registration in Charm. Am I perhaps missing something about how the --usetransform flag is intended to be used?

Thank you!

**Comment by oulap** [comment]
Is the flirt registration matrix in RAS? charm assumes the input is in RAS and converts to LPS (which is what's used internally by the segmentation approach). If you send me the full flirt command you used on ernie I can try to debug this.

**Reply by axthi** [reply, marked as answer]
Hi,

FSL uses its own convention for the registration matrices, which isn't too straightforward. For affine transformations between subject and MNI space, I came up with the following code to get a registration matrix in world-coordinates (part of the old mri2mesh). I think this can be adapted to your needs ("T1fs_nu_conform" and  "$STD_MNI" would be your two images, MNI2conform_12DOF_fsl.mat corresponds to the fsl trafo that you have now, and MNI2conform_12DOF.mat the final matrix that you pass to charm. For that to work, both images have to be stored in LAS format, though, before applying flirt.

echo 'linear registrations to MNI (fsl flirt) ... '
flirt -dof 12 -in $STD_MNI -ref T1fs_nu_conform -out MNI2conform_12DOF -omat MNI2conform_12DOF_fsl.mat

# convert matrices from FSL format to "SPM" format (NOTE: this works only correctly as long as both images are in LAS space)
# the latter trafos from conformed mm world space to MNI mm world space (and inverse)
m=(`fslorient -getqform T1fs_nu_conform`)
cat > qform_conf.mat <<EOF
  ${m[0]} ${m[1]} ${m[2]} ${m[3]}
  ${m[4]} ${m[5]} ${m[6]} ${m[7]}
  ${m[8]} ${m[9]} ${m[10]} ${m[11]}
  ${m[12]} ${m[13]} ${m[14]} ${m[15]}
EOF

m=(`fslorient -getqform $STD_MNI`)
cat > qform_mni.mat <<EOF
  ${m[0]} ${m[1]} ${m[2]} ${m[3]}
  ${m[4]} ${m[5]} ${m[6]} ${m[7]}
  ${m[8]} ${m[9]} ${m[10]} ${m[11]}
  ${m[12]} ${m[13]} ${m[14]} ${m[15]}
EOF

# create internal FSL matrix for MNI template
VOXSIZE_X=`fslval $STD_MNI pixdim1`
VOXSIZE_Y=`fslval $STD_MNI pixdim2`
VOXSIZE_Z=`fslval $STD_MNI pixdim3`
cat > fsl_mni.mat <<EOF
  $VOXSIZE_X     0           0        0
     0        $VOXSIZE_Y     0        0
     0           0       $VOXSIZE_Z   0
     0           0           0        1
EOF

# create internal FSL matrix for conformed image
VOXSIZE_X=`fslval $M2M_DIR/T1fs_nu_conform pixdim1`
VOXSIZE_Y=`fslval $M2M_DIR/T1fs_nu_conform pixdim2`
VOXSIZE_Z=`fslval $M2M_DIR/T1fs_nu_conform pixdim3`
cat > fsl_conf.mat <<EOF
  $VOXSIZE_X        0          0        0
     0          $VOXSIZE_Y     0        0
     0              0      $VOXSIZE_Z   0
     0              0          0        1
EOF

# mm-to-mm mapping from MNI to conformed space (linear):
# Pos_conf = qform_conf*inv(fsl_conf)*R_fsl*fsl_mni*inv(qform_mni)*Pos_MNI
convert_xfm -omat qform_mni_inv.mat -inverse qform_mni.mat
convert_xfm -omat postmat.mat -concat fsl_mni.mat qform_mni_inv.mat

convert_xfm -omat fsl_conf_inv.mat -inverse fsl_conf.mat
convert_xfm -omat premat.mat -concat qform_conf.mat fsl_conf_inv.mat

convert_xfm -omat MNI2conform_12DOF.mat -concat MNI2conform_12DOF_fsl.mat postmat.mat
convert_xfm -omat MNI2conform_12DOF.mat -concat premat.mat MNI2conform_12DOF.mat

**Reply by cascam07** [reply]
Thank you so much! I implemented this solution in a script and am now seeing the expected registration when I use the script's output as the "flirtmat" in my charm call.

**Reply by cascam07** [reply]
@axthi @oulap 
I have a follow up question I'd appreciate input on. 

Would it be possible to apply a nonlinear warp to the template before running charm? I ask because even a very good affine registration between a baby's brain and an adult template is going to have many issues due to major differences in the relative sizing/spacing of tissues. E.g. there is much less space between brain and skull in an infant compared to an adult and infant skulls are overall much thinner even after accounting for head size. So if the template could be pre-warped for each baby, I suspect we would get much better registration and segmentation. However, I don't know how applying such a warp to the template would impact the relationship between it and the two-level mesh atlases. Is this remotely feasible?

Thank you for your thoughts!

**Reply by oulap** [reply]
It's possible but not super straight-forward. You'd need to warp the nodes while making sure the mesh doesn't fold or tear. We can drive the mesh deformation towards some target node positions, which in your case would be the ones after warping. The code behind the link is used to warp a mesh to another mesh (with possibly different connectivity and number of nodes) given the two meshes and a deformation (defined on mesh nodes): 

https://github.com/simnibs/simnibs/blob/master/simnibs/segmentation/samseg/warp_mesh.py

We could probably use that to deform the template meshes to some warped position while making sure that nothing bad happens to the mesh. If you can give me an example warp, e.g., on ernie, I can try to implement a function that applies that to the meshes.
